### ISSUE TITLE
Helm: resource ordering not working

### ISSUE THREAD

**Issue by geekflyer** [pulumi/pulumi-kubernetes#239]
The helm implementation does some ordering in an attempt to deploy resources in a "more likely to work" order as implemented here: https://github.com/pulumi/pulumi-kubernetes/blob/master/sdk/nodejs/helm.ts#L109 .
Unfortunately this doesn't seem to work at all or have any effect on pulumis deploy order in practice. Resources get deployed in completely random order and it hence it fails.

The best way to reproduce this is to attempt to install istio, which has a lot of resources:
1. `helm repo add istio https://istio.io/charts`
2. deploy:
```ts
import { helm, core } from '@pulumi/kubernetes';

const appName = 'istio';

const namespaceName = `${appName}-system`;

const namespace = new core.v1.Namespace(namespaceName, {
  metadata: { name: namespaceName }
});

const chart = new helm.v2.Chart(
  appName,
  {
    repo: 'istio',
    chart: appName,
    namespace: namespaceName,
    version: '1.0.1',
    // for all options check https://github.com/istio/istio/tree/master/install/kubernetes/helm/istio
    values: {}
  },
  { dependsOn: [namespace] }
);
```

**Comment by lukehoban** [comment]
I believe this is more or less the same issue tracked in https://github.com/pulumi/pulumi/issues/1990, though I believe we will need some fix in the pulumi-kubernetes later to address this.

**Comment by geekflyer** [comment]
Just read through #1990 .
Seemed to me the best way to address this would be:
1. Deploy all resources in a chart with max parallelism. Probably this isn't really a proper solution and in many cases like CRDs it will still require multiple `pulumi up` attempts.
2. Infer the `dependsOn` from the rendered manifests. This would be probably the best and most reliable solution but also requires the most implementation work. I believe generally speaking one can analyze / build a dependency graph out of rendered manifests. This unfortunately requires probably a lot of resource kind specific logic to understand their dependencies. It may also not work well to analyze dependencies between custom resources. Nevertheless, once the dependencies between the manifests are understood, the dependency information can be passed along to pulumi engine. For a quick start the "inference" code could just infer a dependency graph based on the `kind` ordering of helmSort. Later on one could make the inference a bit smarter by looking at explicit reference like secret and configmap references in a Deployment spec etc..
As for dependencies of custom resources: I think one way to handle this is to have the ability to override / pass  some dependency information to the Chart or ConfigGroup pulumi resource (i.e. similar to transformations). I'm not sure exactly how the API of passing / overriding dependency info should look like yet (maybe as a starting point also just some array of `kind` ordering), but it's something worth to discuss.

**Comment by hausdorff** [comment]
Helm implements a topological ordering over `kind` rather than a total order over resources (which is what `dependsOn` is). This seems to be much, _much_ simpler to implement, and much closer to what we want anyway. Consider, for example, that Helm mandates that `Service` comes before `Deployment`; if a dependency ordering is captured with `dependsOn`, that pretty much guarantees all charts that have both of those things will fail.

The problem is that I don't believe this _can_ be solved only at the Helm layer, as @lukehoban suggests -- at least not unless we're planning on going the `dependsOn` route, which I think is too complex to be feasible. I'd love to be wrong on that point, because that code would be simpler to write and have a much smaller blast radius.

If we can't come up with a fix purely in the Helm layer, we might instead consider something like this:

- `CollectionComponentResource` to register "scheduling priority" of child resources; resources whose parent is directly the component would then be scheduled for that order.
- Assign priority based on a topological sort of resource `kind`

There are a lot of subtleties in the way that this would have to work, because it's contrary to our current scheduling model. The scheduler needs a way to know when it can start scheduling the resources in the collection, or it could accidentally end up doing some operation only to discover later that there was a higher-priority one. This seems to imply that collections would need to be first-class, and then you have even more questions like, can they be nested, _etc_.

**Comment by geekflyer** [comment]
I haven't really looked in detail at the pulumi helm code, but what I meant with inferring a dependency graph based on `kind` and setting the `dependsOn` could look like this in ts pseudo-code:

```ts
import { groupBy } from 'lodash';

const parsedManifests: Manifest[] = renderAndParseManifestsFromSomewhereEgAHelmChart();

const kindOrder = ['ServiceAccount', 'Deployment', 'Service'];

const manifestsGroupedByKind = groupBy(parsedManifests, manifest => manifest.kind);

let resourcesThatShouldBeCreatedFirst = [];

kindOrder.forEach(kind => {
  const manifestsForKind = manifestsGroupedByKind[kind];
  if (!manifestsForKind) {
    return;
  }
  const resources = manifestsForKind.map(manifest => turnManifestIntoResourceSomeHow(manifest, {dependsOn: resourcesThatShouldBeCreatedFirst}));
  if (resources.length > 0) {
    resourcesThatShouldBeCreatedFirst = resources;
  }
});
```

assuming we have 1 ServiceAccount (A), 2 Deployments (B,C) and 2 Services (D,E) this would create this resource graph:

https://mermaidjs.github.io/mermaid-live-editor/#/view/eyJjb2RlIjoiZ3JhcGggQlRcbkIgLS0-IEFcbkMgLS0-IEFcbkQgLS0-IEJcbkQgLS0-IENcbkUgLS0-IENcbkUgLS0-IEJcbiIsIm1lcm1haWQiOnsidGhlbWUiOiJkZWZhdWx0In19

Is that the same thing as one of the options you're suggesting @hausdorff  ?

**Comment by hausdorff** [comment]
Kind of, except I'm saying that the order can't be strict even among kind. Services come before deployments in helm, so we basically need an order in which to issue the resource operations but without the requirement that they complete.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by geekflyer** [comment]
Well yeah, ideally that would be the case, but this requires more modifications to the core engine I guess and more effort to implement. As a quick fix I think translating the kind order into a strict dependency graph is a good start.
The potential downsides of this approach I see are:
- potentially slow when deploying lots of different kinds, because the engine will await completion of all resources per kind before continuing to deploy the next kind.
- an issue with one upstream resource might prevent deployment of a lot of downstream resources that technically don't actually depend on the upstream resource. However I believe as long as one creates seperate ConfigGroups and Charts for each "logical application unit" this is less of an issue. Like what's the use of a half-broken / half-deployed istio chart anyways?

**Comment by geekflyer** [comment]
Oh, now I'm just realizing what you meant with Services come before Deployments in helm. That indeed is an issue. The reason is probably that some Deployments depend on other Services and their liveness / readyness probes won't even pass before that other Service is up.

**Comment by hausdorff** [comment]
Yes, that's right. There could be circular dependencies, which implies a total order using `depensOn` will not work, and has a very high likelihood of breaking most Charts.

In the extremely short term, I think the best bet is to grant Kubernetes applications "unbounded" parallelism, so that all resources are submitted in parallel, but they don't have to wait for creation to continue. I believe this should work even many very complicated Charts that have this problem, like `kube-prometheus`, which is ~20-30 resources.

This leaves us with Charts in the same class as Istio, where the wheels of this proposal might still fall off.

As I said in https://github.com/pulumi/pulumi/issues/1990, even Helm and `kubectl` do not solve this problem 100% of the time, so I think our actual goal is just to be as good as them. I believe that the "topological sort on `kind`" type solution is pretty much at parity with either of them. I'll chat with @swgillespie today to see what's possible.

**Comment by hausdorff** [comment]
Moving this out to M19. See [this comment](https://github.com/pulumi/pulumi-kubernetes/issues/240#issuecomment-429952763) in #240.

**Comment by hausdorff** [comment]
See also https://github.com/pulumi/pulumi/issues/1743. Solving this issue will probably involve solving that issue.

**Comment by hausdorff** [comment]
After some consideration, we now believe this can be solved with some relatively simple changes in the Kubernetes provider and the Helm SDK. Quoting from https://github.com/pulumi/pulumi/issues/1990#issuecomment-435517052:

> After some consideration, we've agreed to try a simpler approach, which would make Pulumi much less vulnerable to ordering issues, and which seems likely to fix most manifestations of this bug. If we are wrong, we can re-open this bug and consider other options -- our goal will be to implement the simple solution quickly to make sure we have time to come back to this if we want to.
>
> The solution will be occur in the Kubernetes provider, and in the Helm SDK, and involves:
>
> * **Keep Helm's sort order.** Even though this does not guarantee resources will actually be created in that order, in general it increases the likelihood that we will have approximately the correct order.
> * **Set default parallelism to be unbounded.** This is resolved in #2123.
> * **Retry `create` REST requests if they fail or conflict.** We will _not_ retry the await logic. For example, if the API server rejects a resource because of the namespace doesn't exist, or the definition is not found, the REST request will be tried a few more times. This is similar to how `kubectl` handles the conflicts ([_e.g._](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/apply/apply.go#L94).
> * **Add a little extra dependency information in the Helm SDK.** For example, because we know about `Deployment`'s schema, we can make sure that it depends on the `ConfigMap`s and `Secret`s that it references -- usnig a `.get` if they're not created by the Chart.
>
> Because the current fix is likely to exist entirely in the provider and the Helm SDK, I will close this bug, since there is nothing left to do in pulumi/pulumi.
>
> Let me know if you think I missed something.
